File: engine-py/pipeline.py
#!/usr/bin/env python3
"""
Main pipeline orchestrator — subprocess model.

Each parser runs as an isolated subprocess with its own Python interpreter,
completely bypassing the GIL. This mirrors what the Rust orchestrator does
and gives true parallelism for all parser types.

Concurrency (same limits as the Rust orchestrator):
  - Office  : 8 concurrent  (CPU-only, sub-second)
  - Docling : 1 concurrent  (GPU-bound — 2+ causes VRAM OOM)
  - VLM     : 4 concurrent  (I/O-bound, network to vLLM server)

Parser scripts are reused from py-rs-version/parsers/.
pypdfium2 (classification + long-PDF extraction) stays in the main thread
because the C library is not thread-safe.
"""
from __future__ import annotations

import json
import subprocess
import sys
import threading
import time
from concurrent.futures import ThreadPoolExecutor, as_completed
from pathlib import Path

from config import Config
from detector import DocClass, classify, extract_pdf_pages
from parsers import ParseResult

# ── Concurrency gates ─────────────────────────────────────────────────────────

_OFFICE_SLOTS  = threading.Semaphore(8)
_DOCLING_SLOTS = threading.Semaphore(1)
_VLM_SLOTS     = threading.Semaphore(4)

# ── Parser script locations ───────────────────────────────────────────────────

_PARSERS_DIR = Path(__file__).parent.parent / "py-rs-version" / "parsers"

_SCRIPT: dict[DocClass, str] = {
    DocClass.OFFICE:        "office_parser.py",
    DocClass.PDF_SHORT_TEXT: "docling_parser.py",
}
# All other classes → vlm_parser.py


def _semaphore_for(cls: DocClass) -> threading.Semaphore:
    if cls == DocClass.OFFICE:
        return _OFFICE_SLOTS
    if cls == DocClass.PDF_SHORT_TEXT:
        return _DOCLING_SLOTS
    return _VLM_SLOTS


def _script_for(cls: DocClass) -> Path:
    return _PARSERS_DIR / _SCRIPT.get(cls, "vlm_parser.py")


def _vlm_extra(cfg: Config) -> list[str]:
    return [
        "--vlm-url",     cfg.vlm_base_url,
        "--vlm-model",   cfg.vlm_model,
        "--vlm-tokens",  str(cfg.vlm_max_tokens),
        "--vlm-dpi",     str(cfg.vlm_image_dpi),
        "--vlm-workers", str(cfg.vlm_max_workers),
        "--vlm-retries", str(cfg.vlm_retry_attempts),
        "--vlm-backoff", str(cfg.vlm_retry_backoff),
    ]


# ── Per-file subprocess call ──────────────────────────────────────────────────

def _process_file(
    parse_path: Path,    # actual file to parse (may be a temp for long PDFs)
    cls: DocClass,
    original_path: Path, # used for display and output filename
    cfg: Config,
) -> ParseResult:
    t0 = time.time()

    extra = _vlm_extra(cfg) if cls not in (DocClass.OFFICE, DocClass.PDF_SHORT_TEXT) else []
    cmd   = [
        sys.executable, str(_script_for(cls)),
        "--input",      str(parse_path),
        "--output-dir", str(cfg.output_dir),
    ] + extra

    proc    = subprocess.run(cmd, capture_output=True, text=True)
    elapsed = time.time() - t0

    if proc.stderr:
        print(proc.stderr, end="", file=sys.stderr)

    # Parser scripts print one JSON line as the last line of stdout
    json_data: dict = {}
    for line in reversed(proc.stdout.splitlines()):
        line = line.strip()
        if line.startswith("{"):
            try:
                json_data = json.loads(line)
                break
            except json.JSONDecodeError:
                pass

    ok    = json_data.get("ok", False) and proc.returncode == 0
    error = (json_data.get("error") or None) if ok else (
        json_data.get("error") or f"process exited {proc.returncode}"
    )

    result = ParseResult(
        source=original_path,
        parser=json_data.get("parser", "unknown"),
        content="",           # subprocess already wrote the .md file
        page_count=json_data.get("pages"),
        extras={
            "elapsed_s":   round(elapsed, 2),
            "table_count": json_data.get("table_count"),
        },
        error=error,
    )

    if result.ok:
        print(f"    ✓  {original_path.name}  →  {original_path.stem}.md  ({elapsed:.1f}s)")
    else:
        print(f"    ✗  {original_path.name}  ERROR: {result.error}")

    return result


# ── Main entry point ──────────────────────────────────────────────────────────

def run(
    cfg: Config,
    *,
    sort_first: bool = True,
    input_paths: list[Path] | None = None,
) -> list[ParseResult]:
    cfg.ensure_dirs()
    run_start = time.time()

    # ── 1. Sort ───────────────────────────────────────────────────────────────
    if input_paths is None:
        if sort_first:
            from sorter import sort_input
            print("\n── Sorting input files ──────────────────────────────────")
            sorted_map = sort_input(cfg)
            paths = [p for paths in sorted_map.values() for p in paths]
        else:
            paths = [
                f for f in cfg.sorted_dir.rglob("*")
                if f.is_file() and not f.name.startswith("_")
            ]
    else:
        paths = input_paths

    if not paths:
        print("  [pipeline] no files to process")
        return []

    # ── 2. Classify + prepare long PDFs ──────────────────────────────────────
    # Both steps stay in the main thread — pypdfium2 is not thread-safe.
    print(f"\n── Processing {len(paths)} file(s) ─────────────────────────────")
    prepared: list[tuple[Path, DocClass, Path]] = []  # (parse_path, cls, original_path)

    for path in sorted(paths):
        cls = classify(path, cfg)
        print(f"  [{cls.name:18s}]  {path.name}")

        if cls in (DocClass.PDF_LONG_TEXT, DocClass.PDF_LONG_SCAN):
            tmp = cfg.temp_dir / f"_sample_{path.stem}.pdf"
            n   = extract_pdf_pages(path, tmp,
                                    head=cfg.long_head_pages,
                                    tail=cfg.long_tail_pages)
            print(f"    [pipeline] extracted {n} pages "
                  f"(head {cfg.long_head_pages} + tail {cfg.long_tail_pages}) → {tmp.name}")
            prepared.append((tmp, cls, path))

        elif cls == DocClass.UNKNOWN:
            print(f"    [pipeline] WARNING: unreadable — skipping {path.name}")
            prepared.append((path, cls, path))

        else:
            prepared.append((path, cls, path))

    # ── 3. Dispatch parsers concurrently ─────────────────────────────────────
    # Each subprocess has its own Python interpreter — no shared GIL.
    results: list[ParseResult] = []

    def _dispatch(parse_path: Path, cls: DocClass, original_path: Path) -> ParseResult:
        if cls == DocClass.UNKNOWN:
            return ParseResult(
                source=original_path, parser="none", content="",
                error="unreadable PDF (encrypted or corrupted)",
            )
        with _semaphore_for(cls):
            result = _process_file(parse_path, cls, original_path, cfg)
        if parse_path != original_path:
            parse_path.unlink(missing_ok=True)
        return result

    # 13 threads: enough to saturate all semaphore slots simultaneously
    with ThreadPoolExecutor(max_workers=13) as pool:
        futures = {
            pool.submit(_dispatch, pp, cls, op): op
            for pp, cls, op in prepared
        }
        for future in as_completed(futures):
            results.append(future.result())

    # ── 4. Write summary ──────────────────────────────────────────────────────
    summary = []
    ok_count = err_count = 0
    for r in results:
        summary.append({
            "source":      r.source.name,
            "parser":      r.parser,
            "pages":       r.page_count,
            "ok":          r.ok,
            "error":       r.error,
            "elapsed_s":   r.extras.get("elapsed_s"),
            "table_count": r.extras.get("table_count"),
        })
        if r.ok:
            ok_count += 1
        else:
            err_count += 1

    total_elapsed = time.time() - run_start
    minutes, seconds = divmod(total_elapsed, 60)
    elapsed_str = f"{int(minutes)}m {seconds:.1f}s" if minutes else f"{seconds:.1f}s"

    summary_path = cfg.output_dir / "_summary.json"
    summary_path.write_text(
        json.dumps(
            {"total_elapsed_s": round(total_elapsed, 2), "files": summary},
            ensure_ascii=False, indent=2,
        ),
        encoding="utf-8",
    )

    total = len(results)
    print(
        f"\n── Done — {ok_count}/{total} succeeded, {err_count} errors  "
        f"[total: {elapsed_str}]\n"
        f"   Output:  {cfg.output_dir}\n"
        f"   Summary: {summary_path}\n"
    )
    return results

File: py-rs-version/src/pipeline.rs
/// Main pipeline orchestrator.
///
/// Each file is dispatched as an independent tokio task.  Concurrency per
/// parser type is capped by semaphores:
///   - Office (Kreuzberg, CPU-only)  → 8 concurrent
///   - Docling (GPU layout+tables)   → 2 concurrent  (VRAM-bound)
///   - VLM (network to vLLM server)  → 4 concurrent  (server-side limit)
///
/// Long PDFs (>threshold pages) have their head+tail pages extracted to a
/// temp file by the Rust extractor before the Python parser is called.
/// The temp file is deleted after the parser returns.
use std::path::{Path, PathBuf};
use std::sync::Arc;
use std::time::Instant;

use anyhow::Result;
use tokio::process::Command;
use tokio::sync::Semaphore;

use crate::config::Config;
use crate::detector::DocClass;
use crate::extractor::extract_pdf_pages;
use crate::summary::FileResult;

const OFFICE_SLOTS: usize  = 8;
const DOCLING_SLOTS: usize = 1; // GPU-bound: one at a time prevents VRAM exhaustion
const VLM_SLOTS: usize     = 4;

// ── Python subprocess ─────────────────────────────────────────────────────────

async fn spawn_parser(
    python_bin: &str,
    script: &Path,
    input: &Path,
    output_dir: &Path,
    extra: &[(&str, String)],
) -> Result<(bool, serde_json::Value)> {
    let mut cmd = Command::new(python_bin);
    cmd.arg(script)
       .arg("--input").arg(input)
       .arg("--output-dir").arg(output_dir);
    for (flag, val) in extra {
        cmd.arg(flag).arg(val);
    }

    let out = cmd.output().await?;

    // Print parser stderr (progress / warnings) to our stdout
    if !out.stderr.is_empty() {
        eprint!("{}", String::from_utf8_lossy(&out.stderr));
    }

    // Last JSON line on stdout is the structured result
    let stdout = String::from_utf8_lossy(&out.stdout);
    let json_val: serde_json::Value = stdout
        .lines()
        .rev()
        .find(|l| l.trim_start().starts_with('{'))
        .and_then(|l| serde_json::from_str(l).ok())
        .unwrap_or(serde_json::Value::Null);

    Ok((out.status.success(), json_val))
}

fn make_file_result(
    source_name: &str,
    success: bool,
    json: serde_json::Value,
    elapsed_s: f64,
    stderr_err: Option<String>,
) -> FileResult {
    if !success || json.is_null() {
        return FileResult {
            source: source_name.to_string(),
            parser: json.get("parser").and_then(|v| v.as_str()).unwrap_or("error").to_string(),
            pages: None,
            ok: false,
            error: stderr_err.or_else(|| Some("parser process failed or returned no JSON".to_string())),
            elapsed_s: Some(elapsed_s),
            table_count: None,
        };
    }

    FileResult {
        source:      source_name.to_string(),
        parser:      json["parser"].as_str().unwrap_or("").to_string(),
        pages:       json["pages"].as_u64().map(|p| p as usize),
        ok:          json["ok"].as_bool().unwrap_or(false),
        error:       json["error"].as_str().filter(|s| !s.is_empty()).map(String::from),
        elapsed_s:   Some(elapsed_s),
        table_count: json["table_count"].as_u64().map(|t| t as usize),
    }
}

// ── Per-file task ─────────────────────────────────────────────────────────────

async fn process_one(path: PathBuf, cls: DocClass, cfg: Arc<Config>, sem: Arc<Semaphore>) -> FileResult {
    let _permit = sem.acquire().await.unwrap();
    let t0 = Instant::now();

    let source_name = path
        .file_name()
        .unwrap_or_default()
        .to_string_lossy()
        .to_string();

    // Long PDF: extract head+tail pages to a temp file first
    let (parse_path, is_temp) = match cls {
        DocClass::PdfLongText | DocClass::PdfLongScan => {
            let tmp = cfg.temp_dir().join(format!("_sample_{}", source_name));
            match extract_pdf_pages(&path, &tmp, cfg.long_head_pages, cfg.long_tail_pages) {
                Ok(n) => {
                    println!(
                        "    [pipeline] {} — extracted {} pages (head {} + tail {}) → {}",
                        source_name, n, cfg.long_head_pages, cfg.long_tail_pages,
                        tmp.file_name().unwrap_or_default().to_string_lossy()
                    );
                    (tmp, true)
                }
                Err(e) => {
                    eprintln!("    [pipeline] page extraction failed for {}: {}", source_name, e);
                    (path.clone(), false)
                }
            }
        }
        _ => (path.clone(), false),
    };

    let script = cfg.parsers_dir.join(cls.parser_script());

    // VLM args passed only to vlm_parser.py
    let vlm_extra: Vec<(&str, String)> = vec![
        ("--vlm-url",      cfg.vlm_base_url.clone()),
        ("--vlm-model",    cfg.vlm_model.clone()),
        ("--vlm-tokens",   cfg.vlm_max_tokens.to_string()),
        ("--vlm-dpi",      cfg.vlm_image_dpi.to_string()),
        ("--vlm-workers",  cfg.vlm_max_workers.to_string()),
        ("--vlm-retries",  cfg.vlm_retry_attempts.to_string()),
        ("--vlm-backoff",  cfg.vlm_retry_backoff.to_string()),
    ];

    let extra: &[(&str, String)] = match cls {
        DocClass::Office | DocClass::PdfShortText => &[],
        _ => &vlm_extra,
    };

    let (success, json) =
        spawn_parser(&cfg.python_bin, &script, &parse_path, &cfg.output_dir(), extra)
            .await
            .unwrap_or_else(|e| (false, serde_json::Value::String(e.to_string())));

    let elapsed = t0.elapsed().as_secs_f64();

    // Clean up temp file
    if is_temp {
        let _ = std::fs::remove_file(&parse_path);
    }

    let mut result = make_file_result(&source_name, success, json, elapsed, None);
    // Always use original filename in the result (not the temp name)
    result.source = source_name.clone();

    if result.ok {
        println!("    ✓  {}  ({:.1}s)", source_name, elapsed);
    } else {
        println!(
            "    ✗  {}  ERROR: {}",
            source_name,
            result.error.as_deref().unwrap_or("unknown")
        );
    }

    result
}

// ── Public entry point ────────────────────────────────────────────────────────

pub async fn process_all(files: Vec<(PathBuf, DocClass)>, cfg: &Config) -> Result<Vec<FileResult>> {
    let cfg = Arc::new(cfg.clone());

    let office_sem  = Arc::new(Semaphore::new(OFFICE_SLOTS));
    let docling_sem = Arc::new(Semaphore::new(DOCLING_SLOTS));
    let vlm_sem     = Arc::new(Semaphore::new(VLM_SLOTS));

    let mut handles = Vec::with_capacity(files.len());

    for (path, cls) in files {
        let cfg_ref = Arc::clone(&cfg);
        let sem = match cls {
            DocClass::Office       => Arc::clone(&office_sem),
            DocClass::PdfShortText => Arc::clone(&docling_sem),
            _                      => Arc::clone(&vlm_sem),
        };

        handles.push(tokio::spawn(process_one(path, cls, cfg_ref, sem)));
    }

    let mut results = Vec::with_capacity(handles.len());
    for handle in handles {
        results.push(handle.await?);
    }
    Ok(results)
}
